vyper/venom/parser.py

```diff
@@ -0,0 +1,143 @@
+from vyper.venom.context import IRContext
+from vyper.venom.basicblock import (
+    IRLabel,
+    IRVariable,
+    IRLiteral,
+    IROperand,
+    IRInstruction,
+    IRBasicBlock,
+)
+from vyper.venom.function import IRFunction
+from lark import Lark, Transformer
+
+
+VENOM_PARSER = Lark(
+    """
+    %import common.CNAME
+    %import common.DIGIT
+    %import common.LETTER
+    %import common.WS
+    %import common.INT
+
+    start: function* data_section
+    function: "fn" NAME "=>" "{" block* "}"
+    data_section: "data:" call*
+
+    block: NAME ":" (call | assignment)*
+
+    assignment: VAR_IDENT "=" expr
+    expr: call | CONST
+    call: CNAME operands_list
+
+    operands_list: (operand ("," operand)*)?
+
+    operand: VAR_IDENT | CONST | LABEL
+    CONST: INT
+
+    VAR_IDENT: "%" INT (":" INT)?
+    LABEL: "@" NAME
+    NAME: (DIGIT|LETTER|"_")+
+
+    %ignore WS
+    """
+)
+
+
+def set_last_var(fn: IRFunction):
+    for block in fn.get_basic_blocks():
+        output_vars = (
+            instruction.output
+            for instruction in block.instructions
+            if instruction.output is not None
+        )
+        for output_var in output_vars:
+            assert isinstance(output_var, IRVariable)
+            value = output_var.value
+            assert value.startswith("%")
+            value = value.replace("%", "", 1)
+            if not value.isdigit():
+                continue
+            fn.last_variable = max(fn.last_variable, int(value))
+
+
+class VenomTransformer(Transformer):
+    def start(self, children) -> IRContext:
+        ctx = IRContext()
+        funcs = children[:-1]
+        data_section = children[-1]
+        for fn_name, blocks in funcs:
+            fn = ctx.create_function(fn_name)
+            for block_name, instructions in blocks:
+                bb = IRBasicBlock(IRLabel(block_name), fn)
+
+                for instruction in instructions:
+                    bb.insert_instruction(instruction)
+
+                # Manually insert because we need to override function entry
+                fn._basic_block_dict[block_name] = bb
+
+            set_last_var(fn)
+
+        ctx.data_segment = data_section
+
+        ctx.chain_basic_blocks()
```

Comment:
does chain_basic_blocks() need to be called? or should we assume the user input is well-formed?

Reply:
I don't quite understand it but I think it's somewhat required to build the basic CFG and determine whether the CFGs terminate or not?

Reply:
Oh wait, I think I found the issue `"revert"` is not considered a basic block terminator (not in `vyper/venom/basicblock.py:BB_TERMINATORS`) so that was giving me errors.

Reply:
Yes, `revert` is not a terminator. You can have multiple reverts in a block and that is ok.

Reply:
How is revert not a terminator?

Reply:
right now we don't consider `revert` a terminator internally since it can be reordered within a basic block. this also gives us the property that there is a single terminator per basic block which some passes might rely on.

we could probably add it to the list of legal terminators, but it might require special rules in optimization passes; for now we just ensure blocks with reverts in them get terminated with `stop` (which gets stripped out during assembly anyway).

here if we wanted to validate user code, i think we could use a variant of chain_basic_blocks, which adds stop to basic blocks ending with `revert`, but otherwise does not add the stops or chaining jumps. (actually i just checked and we might be able to get rid of the chaining jumps in `chain_basic_blocks()` entirely, but we should revisit that later, it's out-of-scope for this PR).

Reply:
Ok for now I'll continue to call `chain_basic_blocks` then

Reply:
Actually no, to ensure minimal modification from user input I'll only add `stop` to blocks that *do not* terminate and contain a revert

Reply:
Thinking about it, it doesn't really make sense to consider `revert` non-terminal, and leads to weird edge cases like this one.

Yes you can rearrange it anywhere but that's more a side-effect of general instruction reordering, `revert` is not unique in being able to be put anywhere in a basic block.